### PR TITLE
WIP refactoring

### PR DESCRIPTION
**Pull request by iamjasonaevans**
Refactored the code to include:
- Use contracts for properties and 
- Extracted method for getting the current user
- Extracted method for getting the todo list
- Now using Operation::Model
- Updated spec

### CODE REVIEW

**Comment by kennethcalamay** [comment]
Hi Mr. Evans, thanks for taking the time to work on our coding challenge. I'm KL of serviceseeking.com.au and I'll be looking at your submission. :-)

File: app/concepts/todo/create.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 class Todo::Create < Trailblazer::Operation
+  include Model
+  model Todo, :create
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb]: Nice! It appears that you understand how to use the `Model` mixin. 👏

File: app/concepts/todo/create.rb [hunk 2]
```diff
@@ -10,41 +12,46 @@ class Todo::Create < Trailblazer::Operation
   #         create a default `todo_list`
   #         associate `todo_list` with `user`
   #         associate `todo` with `todo_list`
+
+  contract do
+    property :title, validates: { presence: true }
+    property :description
+    property :todo_list_id
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb]: Use of a `contract` object! 👍

File: app/concepts/todo/create.rb [hunk 3]
```diff
@@ -10,41 +12,46 @@ class Todo::Create < Trailblazer::Operation
...
-      # assign the todo_list to the user
-      todo_list.user = current_user
-      todo_list.save!
-    else # i.e. if params[:todo_list_id] is not nil
-      todo_list = TodoList.find(params[:todo_list_id])
+    # validate and save 
+    validate(params) do
+      contract.save
     end
+  end
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb]: It's good to see that you've separated the concern of creating a `todo_list` and a `user` from the main concern of creating a `todo`.

**Comment by kennethcalamay** [comment]
This is a good start. How do you feel about refactoring it further. i.e. fully separating the concerns of creating a `user` and creating a `todo_list` into different operations. Also, it will be nice to see exhaustive specs for each of the resulting operations.

**Comment by kennethcalamay** [comment]
I look forward to your next commits sir. Thank you.

**Comment by kennethcalamay** [comment]
Wow, you certainly went overboard (in a good way) sir. It's impressive to say the least. The code can be improved in a lot of ways, but it's understandable since it will come with experience in Trailblazer. Also, OOD can be applied in lots of areas, but I'm pretty sure it's just a review away from you. I will recommend you for a chat with our senior dev in Sydney, one of the founders, himself. Thanks for taking the time to do our coding challenge.

File: app/concepts/todo/cell.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    render
+  end
+
+  def edit
+    render :form
+  end
+
+  private
+
+  include ActionView::Helpers::FormHelper
```
Comment (kennethcalamay) [inline app/concepts/todo/cell.rb]: This is the second time this was included.

File: app/concepts/session/operation.rb [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+    # hash the password
+    def password=(password)
+      @model.password_hash = Password.create(password)
+    end
+  end
+
+  class Delete < Trailblazer::Operation
+    include Model
+    include Dispatch
```
Comment (kennethcalamay) [inline app/concepts/session/operation.rb]: Not sure why dispatch was included.

File: app/concepts/session/operation.rb [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    model User, :find
```
Comment (kennethcalamay) [inline app/concepts/session/operation.rb:94]: Model actions can only either be `:create`, or `:update`

File: app/concepts/todo/operation.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
+class Todo::Create < Trailblazer::Operation
+  include Model
+  model Todo, :create
+
+  contract do
+    property :title, validates: { presence: true }
+    property :description
+    property :todo_list_id
+    property :completed
+  end
```
Comment (kennethcalamay) [inline app/concepts/todo/operation.rb:10]: 👍

File: app/concepts/todo/operation.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  def process(params)
+    # validate and save
```
Comment (kennethcalamay) [inline app/concepts/todo/operation.rb]: Comment not needed.

File: app/concepts/todo/operation.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    validate(params[:todo]) do
+      contract.save
+    end
+  end
+
+end
```
Comment (kennethcalamay) [inline app/concepts/todo/operation.rb:25]: Looks exactly like Todo::Create, might help to just inherit from it and set

```
action :update
```

File: app/concepts/todo/operation.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class Todo::Delete < Trailblazer::Operation
+  include Model
+  model Todo, :find
```
Comment (kennethcalamay) [inline app/concepts/todo/operation.rb:30]: This is the nth time I saw `:find`, does the newest version of Trailblazer make use of it?

Comment (iamjasonaevans) [inline app/concepts/todo/operation.rb:30]: `:find` will execute Model.find(param[:id]), http://trailblazer.to/gems/operation/model.html

File: app/concepts/session/operation.rb (hunk 5 above)
Comment (iamjasonaevans) [inline app/concepts/session/operation.rb]: Yes, this is redundant. I was going to implement a callback, this code can be removed.

File: app/concepts/todo/cell.rb (hunk 4 above)
Comment (iamjasonaevans) [inline app/concepts/todo/cell.rb]: Will remove

File: app/concepts/session/operation.rb (hunk 5 above)
Comment (kennethcalamay) [inline app/concepts/session/operation.rb]: Good to know that you know about callbacks though.

File: app/concepts/todo/operation.rb (hunk 10 above)
Comment (kennethcalamay) [inline app/concepts/todo/operation.rb:30]: I see. I guess my trailblazer knowledge is dated. Thank you for the link.

File: app/concepts/todo/operation.rb (hunk 9 above)
Comment (iamjasonaevans) [inline app/concepts/todo/operation.rb:25]: That works. How about we go one step further and `create` and `update` both inherit from a base class, there seems to be a more a "is a" relationship with a base class? That way we are inheriting the shared `contract` and super of `process` each subclass could then have it's own extended logic potentially saving a refactor later as Service Objects have the potential for complex business logic/rules.

Comment (iamjasonaevans) [inline app/concepts/todo/operation.rb]: `class Todo::BaseClass < Trailblazer::Operation
  include Model
  model Todo

  contract do
    property :title, validates: { presence: true }
    property :description
    property :todo_list_id
    property :completed
  end

  def process(params)
    validate(params[:todo]) do
      contract.save
    end
  end
end

class Todo::Create <  Todo::BaseClass
  action :create
end

class Todo::Update < Todo::BaseClass
  action :update
end`

Comment (kennethcalamay) [inline app/concepts/todo/operation.rb:25]: We could. This works for now. Kudos to you for thinking that way. I'm not surprised though.

**Pull request closed without merging**